internal/olm/operator/catalog_source.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package olm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"strings"
+
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	olmclient "github.com/operator-framework/operator-sdk/internal/olm/client"
+	internalregistry "github.com/operator-framework/operator-sdk/internal/olm/operator/internal"
+
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// CatalogResources holds resources for handling CatalogSource creation
+type CatalogResources struct {
+	namespace       string
+	registryPodName string
+	registryPodAddr string
+
+	catalogsourceObj []runtime.Object
+}
+
+// initialize will initialize catalog resources with values from
+// a registry pod and validate the CatalogResources struct
+func initialize(rp *internalregistry.RegistryPod) (*CatalogResources, error) {
+	if rp == nil {
+		return nil, fmt.Errorf("registry pod is not initialized")
+	}
+
+	podName, err := rp.GetRegistryPodName()
+	if err != nil {
+		return nil, fmt.Errorf("error in getting registry pod name: %v", err)
+	}
+
+	podAddr := internalregistry.GetRegistryServiceAddr(podName, rp.Namespace)
+
+	cr := &CatalogResources{
+		namespace:       rp.Namespace,
+		registryPodName: podName,
+		registryPodAddr: podAddr,
+	}
+
+	// validate the CatalogResources struct and ensure required fields are set
+	if err := cr.validate(); err != nil {
+		return nil, fmt.Errorf("error in validating CatalogResources struct: %v", err)
+	}
```

Comment:
IMO it is not required. why we need to check it? what would be the valid scenario that it would not be created?

Reply:
The scenario is that if we just create a `RegistryPod` without using `NewRegistryPod` function , it's likely that some of the fields might be empty or not set based on how the caller sets it. This method here should validate that the fields such as name, or namespace are not empty.

Reply:
My suggestion is that would be checked here https://github.com/operator-framework/operator-sdk/pull/3467/files#r457750854 instead of creating a new func.

Reply:
We are only checking if rp is nil and the rp.Pod is nil. We still need a validate method to ensure that the catalog source name, namespace and the address are not empty.